x/stake/keeper/slash.go

```diff
@@ -4,56 +4,212 @@ import (
 	"fmt"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	types "github.com/cosmos/cosmos-sdk/x/stake/types"
 	"github.com/tendermint/tendermint/crypto"
 )
 
-// NOTE the current slash functionality doesn't take into consideration unbonding/rebonding records
-//      or the time of breach. This will be updated in slashing v2
-// slash a validator
-func (k Keeper) Slash(ctx sdk.Context, pubkey crypto.PubKey, height int64, fraction sdk.Rat) {
+// slash an unbonding delegation and update the pool
+func (k Keeper) slashUnbondingDelegation(ctx sdk.Context, unbondingDelegation types.UnbondingDelegation, infractionHeight int64, fraction sdk.Rat) (slashAmount sdk.Int) {
+	now := ctx.BlockHeader().Time
+
+	// If unbonding started before this height, stake didn't contribute to infraction
+	if unbondingDelegation.CreationHeight < infractionHeight {
+		return sdk.ZeroInt()
+	}
+
+	if unbondingDelegation.MinTime < now {
+		// Unbonding delegation no longer eligible for slashing, skip it
+		// TODO Settle and delete it automatically?
+		return sdk.ZeroInt()
+	}
+
+	// Calculate slash amount proportional to stake contributing to infraction
+	slashAmount = sdk.NewRatFromInt(unbondingDelegation.InitialBalance.Amount, sdk.OneInt()).Mul(fraction).EvaluateInt()
+
+	// Don't slash more tokens than held
+	unbondingSlashAmount := slashAmount
+	if unbondingSlashAmount.GT(unbondingDelegation.Balance.Amount) {
+		unbondingSlashAmount = unbondingDelegation.Balance.Amount
+	}
+
+	// Update unbonding delegation if necessary
+	if !unbondingSlashAmount.IsZero() {
+		unbondingDelegation.Balance.Amount = unbondingDelegation.Balance.Amount.Sub(unbondingSlashAmount)
+		k.SetUnbondingDelegation(ctx, unbondingDelegation)
+		pool := k.GetPool(ctx)
+		// Burn loose tokens
+		// Ref https://github.com/cosmos/cosmos-sdk/pull/1278#discussion_r198657760
+		pool.LooseTokens -= slashAmount.Int64()
+		k.SetPool(ctx, pool)
+	}
+
+	return
+}
+
+// slash a redelegation and update the pool
+func (k Keeper) slashRedelegation(ctx sdk.Context, validator types.Validator, redelegation types.Redelegation, infractionHeight int64, fraction sdk.Rat) (slashAmount sdk.Int) {
```

Comment:
Can fraction be renamed to `slashFactor`

Reply:
Changed to `slashFraction`.